syz-manager/covfilter.go

```diff
@@ -0,0 +1,215 @@
+// Copyright 2020 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package main
+
+import (
+	"encoding/binary"
+	"fmt"
+	"os"
+	"regexp"
+
+	"github.com/google/syzkaller/pkg/log"
+	"github.com/google/syzkaller/pkg/mgrconfig"
+	"github.com/google/syzkaller/sys/targets"
+)
+
+type CoverFilter struct {
+	pcStart     uint32
+	pcSize      uint32
+	pcEnd       uint32
+	weightedPCs map[uint32]float32
+
+	bitmapFilename string
+	target         *targets.Target
+}
+
+func createCoverageFilter(cfg *mgrconfig.Config, target *targets.Target) (covFilterFilename string, err error) {
+	files := cfg.CovFilter.Files
+	funcs := cfg.CovFilter.Functions
+	rawPCs := cfg.CovFilter.RawPCs
+	if len(files) == 0 && len(funcs) == 0 && len(rawPCs) == 0 {
+		return "", nil
+	}
+	filesRegexp, err := getRegexps(files)
+	if err != nil {
+		return "", err
+	}
+	funcsRegexp, err := getRegexps(funcs)
+	if err != nil {
+		return "", err
+	}
+	if len(filesRegexp) > 0 || len(funcsRegexp) > 0 {
+		log.Logf(0, "initialize coverage information...")
+		if err = initCover(target, cfg.KernelObj, cfg.KernelSrc, cfg.KernelBuildSrc); err != nil {
+			return "", err
+		}
+	}
+
+	covFilter := CoverFilter{
+		weightedPCs:    make(map[uint32]float32),
+		target:         target,
+		bitmapFilename: cfg.Workdir + "/" + "syz-cover-bitmap",
+	}
+	err = covFilter.initWeightedPCs(filesRegexp, funcsRegexp, rawPCs)
+	if err != nil {
+		return "", err
+	}
+	err = covFilter.createBitmap()
+	if err != nil {
+		return "", err
+	}
+	return covFilter.bitmapFilename, nil
+}
+
+func (covFilter *CoverFilter) initWeightedPCs(filesRegexp, funcsRegexp []regexp.Regexp, rawPCsFiles []string) error {
+	for _, f := range rawPCsFiles {
+		rawFile, err := os.Open(f)
+		if err != nil {
+			return fmt.Errorf("failed to open raw PCs file: %v", err)
+		}
+		for {
+			var encode uint64
+			_, err := fmt.Fscanf(rawFile, "0x%x\n", &encode)
+			if err != nil {
+				break
+			}
+			pc := uint32(encode & 0xffffffff)
+			weight := float32((encode >> 32) & 0xffff)
+			covFilter.weightedPCs[pc] = weight
+		}
+		rawFile.Close()
+	}
+
+	enabledFuncs := make(map[string]bool)
+	enabledFiles := make(map[string]bool)
+	if len(filesRegexp) > 0 || len(funcsRegexp) > 0 {
+		if reportGenerator == nil {
+			return fmt.Errorf("reportGenerator used without initialization")
+		}
+		pcs := make([]uint64, 0)
+		symNames, cuNames, symPCs := reportGenerator.GetSymbolsInfo()
```

Comment:
It feels that this call is better located next to the initCover call above. It would make the reportGenerator check above not necessary and make code that does one thing located together.
Something along the following lines looks more coherent:
```
	covFilter := CoverFilter{
		...
	}
	if len(filesRegexp) > 0 || len(funcsRegexp) > 0 {
		log.Logf(0, "initialize coverage information...")
		if err = initCover(target, cfg.KernelObj, cfg.KernelSrc, cfg.KernelBuildSrc); err != nil {
			return "", err
		}
		symbols := reportGenerator.GetSymbolsInfo()
		if err := covFilter.initFilesFuncs(filesRegexp, funcsRegexp, symbols); err != nil {
			return "", err
		}
	}
```
And initialization of raw pcs can be moved to a separate function, it does not seem to be much related to funcs/files.